028-service-binding.md

```diff
@@ -0,0 +1,340 @@
+
+Contents:
+
+ - [Current problems with integrating Strimzi and a Service Binding Operator](#current-problems-with-integrating-strimzi-and-a-service-binding-operator)
+ - [Proposal](#proposal)
+ - [Overview of Service Binding specification](#overview-of-service-binding-specification)
+ - [Connecting to Strimzi](#connecting-to-strimzi)
+
+## Current problems with integrating Strimzi and a Service Binding Operator
+
+At a high level these are the current problems that make integrating Strimzi with a Service Binding Operator hard:
+
+1. Bootstrap Server information is not contained in any secret
+2. Information is spread across multiple `Secret` resources
+3. The user has to determine which `KafkaUser` can access which listener
+
+## Proposal
+
+I do not see a single approach that easily solves this problem. I have detailed a few below to enable the Strimzi and Service Binding specification communities to discuss how these two technologies should best be integrated.
+
+Proposed approaches:
+1. Application uses `KafkaUser` to get credentials but does not use the service binding mechanism for cluster certificate or bootstrap server
+2. Application uses multiple `ServiceBinding` resources to get information
+3. Application binds to the `KafkaUser` resource which contains all required information
+4. Application can bind to a single custom resource to get all information
+
+### 1 - Application binds to KafkaUser only
+
+In this approach the application creates a `ServiceBinding` for a specific `KafkaUser` CR. The credential information is bound to the application and the cluster certificate and bootstrap server address is obtained by the application in another way.
+
+Required changes:
+
+ - Update the `KafkaUser` CR to include a `status.binding` field where the `name` matches the current field `secret` in the status
+ - Update the secret created for the `KafkaUser` to include the username when the `KafkaUser` is using type SASL SCRAM
+
+Considerations:
+
+ - In this approach a lot of work is still left up to the user to discover the cluster certificate and the bootstrap server address to use.
+
+### 2 - Application uses multiple ServiceBinding resources
+
+In this approach the application creates two `ServiceBinding` resources, one that will bind to the `Kafka` CR and one that will bind to the `KafkaUser` CR.
+
+Required changes:
+
+- Update the `KafkaUser` CR to include a `status.binding` field where the `name` matches the current field `secret` in the status
+- Update the secret created for the `KafkaUser` to include the username when the `KafkaUser` is using type SASL SCRAM
+- Update the `Kafka` CR to include a `status.binding` field, where the `name` refers to a new secret containing both the cluster certificate and bootstrapServers
+- Add a new secret that contains the bootstrap servers for each listener and the cluster certificate information
```

Comment:
How does the user choose the listener? Will they need to use the `env` section in the `ServiceBinding` resource?

Reply:
As noted in the considerations, this would rely on the application developer writing custom logic to grab the listener they want from the ones that were mounted into the app using the secret.

I would expect we would end up with something like:
```
data:
  listeners:
    - name: internal
       bootstrapServers: kafka.svc:9092
    - name: external
       bootstrapServers: myhost.com:443
```

And then the app gets that mounted as a file and can find the one they want. Not ideal, but it's more useful than what we have today for service binding

Reply:
for the record, k8s secret data are flat, i.e. key values are just simple strings, and if the values are anyhow structured, the application should need to parse them appropriately. 

BTW, does the app really need to have a few bootstrap servers to chose from? Should not we simply project just enough connection data (e.g. single bootstrap server and certificates if needed) and the app just take it and try to connect? Similar to what we do in jdbc world?

Reply:
> for the record, k8s secret data are flat

Ah I hadn't realised that the secret data had to be flat, so I guess any structure would have to be encapsulated in a string.

> BTW, does the app really need to have a few bootstrap servers to chose from?

Strimzi allows users to expose multiple different endpoints for applications to connect to. The user can configure endpoints that are only accessible within the cluster, or accessible externally, or have different security mechanisms. There is no way for Strimzi to know which listener the application will want to use. If there is only one internal listener then Strimzi could guess it is that one, but there could be more than one.

As noted in the considerations, approach number 2 is not ideal for this very reason - it still leaves the application work to do to figure out which endpoint to use. Approaches 3 and 4 both include options that would allow Strimzi to give the application only a single bootstrap server to call.

Reply:
I've updated the examples

Reply:
I'm going to mark this conversation as resolved since this option has been rejected.